// wow-armory.js/armory.js
const axios = require('axios');
const qs = require('qs');

class Armory {

    /**
     * Handles the class assignments for the Battle.net environment variables.
     *
     * @param bnetId
     * @param bnetSecret
     * @param bnetRegion
     */
    constructor(bnetId, bnetSecret, bnetRegion) {
        this.BNET_ID = bnetId;
        this.BNET_SECRET = bnetSecret;
        this.BNET_REGION = bnetRegion;
    }

    /**
     *
     * @param region
     * @param server
     * @param character
     * @param params
     * @returns {Promise<Error>}
     */
    async getCharacter(region, server, character, params = 'achievements,mounts,pets,titles,reputation,progression,items,stats,professions,pvp') {
        try {
            let accessToken = await this.getAccessToken();

            return await axios({
                method: 'get',
                url: `https://${region}.api.blizzard.com/wow/character/${server}/${character}?fields=${params}&locale=en_US&access_token=${accessToken}`
            });
        } catch (err) {
            console.log(err);
            return new Error(err);
        }
    }

    async getAccessToken() {
        const data = {"grant_type": "client_credentials"};

        try {
            let response = await axios({
                method: 'post',
                url: `https://${this.BNET_REGION}.battle.net/oauth/token`,
                withCredentials: true,
                auth: {
                    username: this.BNET_ID,
                    password: this.BNET_SECRET
                },
                headers: {
                    'Content-Type': 'application/x-www-form-urlencoded',
                },
                data: qs.stringify(data)
            });

            return response.data.access_token;
        } catch (err) {
            console.log(err);
            return new Error(err);
        }
    }
}

module.exports = Armory;
